File: day05/part2.pl
#!/usr/bin/env perl
use List::Util qw(min max);
use strict;

my %CELLS;
while (<>) {
  /^(\d+),(\d+) -> (\d+),(\d+)$/;
  if ($1 == $3 || $2 == $4) {
    for my $x (min($1,$3)..max($1,$3)) {
      for my $y (min($2,$4)..max($2,$4)) {
        $CELLS{int($x)}{int($y)} += 1;
      }
    }
  }
  else
  {
    if ($1 < $3) {
      if ($2 < $4) {
        for my $x ($1..$3) {
          $CELLS{int($x)}{int($2+($x-$1))} += 1;
        }
      } else {
        for my $x ($1..$3) {
          $CELLS{int($x)}{int($2-($x-$1))} += 1;
        }
      }
    } else {
      if ($2 < $4) {
        for my $x ($3..$1) {
          $CELLS{int($x)}{int($4-($x-$3))} += 1;
        }
      } else {
        for my $x ($3..$1) {
          $CELLS{int($x)}{int($4+($x-$3))} += 1;
        }
      }
    }
  }
}

my $count;
foreach my $x (keys %CELLS) {
  foreach my $y (keys %{%CELLS{$x}}) {
    if ($CELLS{$x}{$y} > 1) {
      $count += 1;
    }
  }
}
print "$count\n";

File: day09/part2.php
#!/usr/bin/env php
<?php

$CELLS = array();
while ($line = fgets(STDIN)) {
	$CELLS[] = str_split(trim($line));
}
$H = count($CELLS);
$W = count($CELLS[0]);

for ($x = 0; $x < $W; $x++) {
	for ($y = 0; $y < $H; $y++) {
		$CELLS[$y][$x] = intval($CELLS[$y][$x]);
	}
}

function value($x, $y) {
	global $H, $W, $CELLS;
	if ($x >= 0 && $x < $W && $y >= 0 && $y < $H) {
		return $CELLS[$y][$x];
	}
	return 9;
}

function neighbour_values($x, $y) {
	return [
		value($x+1, $y),
		value($x-1, $y),
		value($x, $y+1),
		value($x, $y-1),
	];
}

function add_neighbour_basin(&$todo, &$visited, $x, $y) {
	global $W;
	$N = $y * $W + $x;
	if (!array_key_exists($N, $visited) && value($x, $y) < 9) {
		$todo[] = [$x, $y];
	}
}

function find_basin($x, $y) {
	$visited = [];
	$todo = [[$x, $y]];

	$size = 0;
	global $H, $W, $CELLS;
	while (count($todo) > 0) {
		[$x, $y] = array_pop($todo);
		$N = $y * $W + $x;
		if (!array_key_exists($N, $visited)) {
			$visited[$N] = true;
			$size += 1;
			add_neighbour_basin($todo, $visited, $x+1, $y);
			add_neighbour_basin($todo, $visited, $x-1, $y);
			add_neighbour_basin($todo, $visited, $x, $y+1);
			add_neighbour_basin($todo, $visited, $x, $y-1);
		}
	}
	return $size;
}

$BASINS = [];
for ($x = 0; $x < $W; $x++) {
	for ($y = 0; $y < $H; $y++) {
		if ($CELLS[$y][$x] < min(neighbour_values($x, $y))) {
			$BASINS[] = find_basin($x, $y);
		}
	}
}
rsort($BASINS);
$result = $BASINS[0] * $BASINS[1] * $BASINS[2];
echo "$result";
